tests/lax_control_flow_test.py

```diff
@@ -992,6 +992,78 @@ def fun(carry, _):
 
     api.grad(lambda x: jit_run_scan(x))(0.)  # doesn't crash
 
+  def test_root_scalar(self):
+
+    def scalar_solve(f, y):
+      return y / f(1.0)
+
+    def binary_search(func, x0, low=0.0, high=100.0, tolerance=1e-6):
+      del x0  # unused
+
+      def cond(state):
+        low, high = state
+        return high - low > tolerance
+
+      def body(state):
+        low, high = state
+        midpoint = 0.5 * (low + high)
+        update_upper = func(midpoint) > 0
+        low = np.where(update_upper, low, midpoint)
+        high = np.where(update_upper, midpoint, high)
+        return (low, high)
+
+      solution, _ = lax.while_loop(cond, body, (low, high))
+      return solution
+
+    def sqrt_cubed(x, tangent_solve=scalar_solve):
+      f = lambda y: y ** 2 - x ** 3
+      return lax.root(f, 0.0, binary_search, tangent_solve)
+
+    value, grad = api.value_and_grad(sqrt_cubed)(5.0)
+    self.assertAllClose(value, 5 ** 1.5, check_dtypes=False)
+    self.assertAllClose(grad, api.grad(pow)(5.0, 1.5), check_dtypes=False)
+
+    jtu.check_grads(sqrt_cubed, (5.0,), order=2, rtol=1e-3)
+
+    inputs = np.array([4.0, 5.0])
+    results = api.vmap(sqrt_cubed)(inputs)
+    self.assertAllClose(results, inputs ** 1.5, check_dtypes=False)
+
+    results = api.jit(sqrt_cubed)(5.0)
+    self.assertAllClose(results, 5.0 ** 1.5, check_dtypes=False)
+
+  def test_root_vector(self):
+
+    def oracle(func, x0):
+      del func  # unused
+      return x0
+
+    def vector_solve(f, y):
+      return np.linalg.solve(api.jacobian(f)(y), y)
+
+    def linear_solve(a, b):
+      f = lambda y: np.dot(a, y) - b
+      x0 = np.linalg.solve(a, b)
+      return lax.root(f, x0, oracle, vector_solve)
```

Comment:
Looking at this example, it occurs to me that a better API for `lax.root` is probably `root(f, solution, tangent_solve)`. We don't need to include `solve` or `initial_guess`: the `solution` itself suffices for knowing the structure of the output, and all we were using `solve` for was to evaluate `solution = solve(f, initial_guess)`.

Reply:
I gave this a shot in the last commit. It looks a lot cleaner, but now my binary search example fails with `NotImplementedError: Forward-mode differentiation rule for 'while' not implemented`. I would be interested to hear any ideas for working around this...

Reply:
I think that's what we avoided by computing `solution = solve(f, initial_guess)` inside the primitive, i.e. with the old API. With this API the root-finding computation itself is outside the primitive, which means we try to differentiate through it.

I think it's better to go back to the old API, basically `solution = root(f, initial_guess)` (plus parameterizing by `solve` and `tangent_solve`). WDYT?